File: packages/catalyst-llm-sdk/src/agent/events.ts
/**
 * TypeScript mirror of catalyst-langgraph's AgentEvent union.
 *
 * The Python source of truth lives at
 *   packages/catalyst-langgraph/src/catalyst_langgraph/events.py
 *
 * Each event in the union carries a discriminating `type` field so
 * consumers can switch over it (or filter via `event.type === "token"`)
 * without runtime tag inspection. Keep this file in lock-step with
 * the Python schema — adding an event there but not here means the
 * UI silently drops events; vice-versa means we'll log "unknown type"
 * warnings until the backend ships matching code.
 */

/**
 * Mixin for events that can be nested inside a tool execution. When
 * `owner_tool_id` is set, the event was produced INSIDE that tool
 * call (e.g. a council member's tokens while `research` is running);
 * the UI routes such events into the parent tool card's expandable
 * "reasoning" section rather than accumulating them on the
 * assistant turn's main content. When null/undefined, the event came
 * from the parent agent and renders inline as before.
 */
export interface Nestable {
  owner_tool_id?: string | null;
}

export interface RunStarted {
  type: "run_started";
  run_id: string;
  model: string;
}

export interface Token extends Nestable {
  type: "token";
  content: string;
}

/** Reasoning-trace deltas (e.g. `<think>` blocks from r1-style models). */
export interface Reasoning extends Nestable {
  type: "reasoning";
  content: string;
}

export interface ToolCallStart extends Nestable {
  type: "tool_call_start";
  id: string;
  name: string;
  args: Record<string, unknown>;
}

export interface ToolCallEnd extends Nestable {
  type: "tool_call_end";
  id: string;
  result?: unknown;
  error?: string;
  duration_ms: number;
}

export interface Iteration extends Nestable {
  type: "iteration";
  n: number;
}

export interface MessageDone {
  type: "message_done";
  finish_reason?: string;
  usage?: Record<string, unknown>;
}

export interface ErrorEvent {
  type: "error";
  message: string;
}

export type AgentEvent =
  | RunStarted
  | Token
  | Reasoning
  | ToolCallStart
  | ToolCallEnd
  | Iteration
  | MessageDone
  | ErrorEvent;

/** All event-type discriminators in one place, useful for exhaustiveness checks. */
export const AGENT_EVENT_TYPES = [
  "run_started",
  "token",
  "reasoning",
  "tool_call_start",
  "tool_call_end",
  "iteration",
  "message_done",
  "error",
] as const;
export type AgentEventType = (typeof AGENT_EVENT_TYPES)[number];

/** Body shape for POST /api/chat/stream. */
export interface ChatStreamRequest {
  model: string;
  messages: Array<{
    role: "system" | "user" | "assistant" | "tool";
    content: string;
    tool_call_id?: string;
    [key: string]: unknown;
  }>;
  system_prompt?: string;
  /** Tool names the agent may dispatch (must be in /api/tools). */
  tools?: string[];
  /** Sampling params: temperature, max_tokens, top_p, reasoning_effort, … */
  params?: Record<string, unknown>;
  /**
   * Per-Agent overrides for tunables advertised on GET /api/agents.
   * Shape: `{ "main": { "recursion_limit": 30 }, "research": { "model": "..." } }`.
   * Falls back to defaults / env vars when absent.
   */
  agent_config?: Record<string, Record<string, unknown>>;
}

// ─── Agent registry types (mirror /api/agents schema) ─────────────────

/** One node in an Agent's topology graph. */
export interface AgentTopologyNode {
  id: string;
  type: "start" | "end" | "agent" | "tools";
}

/** One edge between two topology nodes. */
export interface AgentTopologyEdge {
  source: string;
  target: string;
  conditional: boolean;
}

/** Static topology snapshot for the Engine tab to render. */
export interface AgentTopology {
  nodes: AgentTopologyNode[];
  edges: AgentTopologyEdge[];
}

/**
 * UI hints attached to a property in the JSON Schema returned by
 * GET /api/agents. Lives at `properties.<field>.ui` and drives the
 * AgentConfigForm's widget choice:
 *   - `widget: "model"`     → ModelSelector dropdown (populated from /api/models)
 *   - `widget: "textarea"`  → multi-line Textarea
 *   - undefined / no widget → renderer picks by JSON Schema `type`:
 *                              number → Slider, string → Input,
 *                              boolean → Switch, enum → Select
 * `step` and `secret` are passed verbatim to the chosen widget.
 */
export interface AgentFieldUiHints {
  widget?: "model" | "textarea";
  step?: number;
  secret?: boolean;
}

/**
 * One property entry within an Agent's JSON Schema. Mirrors the
 * subset of JSON Schema (draft 2020-12) that the backend's
 * Pydantic-derived schemas emit. The `ui` extension key is ours.
 */
export interface AgentFieldSchema {
  type: "string" | "number" | "integer" | "boolean";
  title?: string;
  description?: string;
  default?: unknown;
  minimum?: number;
  maximum?: number;
  exclusiveMinimum?: number;
  exclusiveMaximum?: number;
  enum?: string[];
  ui?: AgentFieldUiHints;
}

/**
 * JSON Schema for an Agent's config_model — exactly what
 * Pydantic's `model.model_json_schema()` produces, with our `ui`
 * extension hints sprinkled onto individual properties.
 */
export interface AgentConfigSchema {
  type: "object";
  title?: string;
  description?: string;
  properties?: Record<string, AgentFieldSchema>;
  required?: string[];
}

/** One Agent in the registry — everything the Engine tab needs to render it. */
export interface AgentDescriptor {
  id: string;
  description: string;
  tools: string[];
  topology: AgentTopology;
  config_schema: AgentConfigSchema;
}

/** Response shape for GET /api/agents. */
export interface ListAgentsResponse {
  agents: AgentDescriptor[];
}

File: packages/catalyst-langgraph/src/catalyst_langgraph/events.py
"""Typed event schema for the /api/chat/stream SSE channel.

The UI consumes these as a discriminated union keyed on `type`. They're
a translation of LangGraph's `astream_events(version="v2")` firehose
into a small, stable shape the playground (and any future UI) can rely
on without coupling to LangGraph internals.

Add a new event type here when the UI needs to react to something new.
Resist the urge to forward LangGraph's raw events — the whole point of
this layer is that LangGraph internals can change without UI churn.
"""
from __future__ import annotations

from typing import Any, Literal, Optional, Union

from pydantic import BaseModel, Field


class RunStarted(BaseModel):
    type: Literal["run_started"] = "run_started"
    run_id: str
    model: str


# Optional attribution shared by every "nestable" event below.
#
# When `owner_tool_id` is set, this event was produced INSIDE the
# execution of the named tool call (e.g. the council members'
# tokens while `research` runs). The UI uses this to route the
# event into that tool's expandable "reasoning" section in the
# ToolCallCard rather than dumping it into the parent's chat bubble.
# When None / omitted, the event came from the parent agent itself
# and renders inline as before. Field lives at the event level (not
# in a wrapper) so existing UIs keep working — they can just ignore
# the field and behave as today.
class _Nestable(BaseModel):
    owner_tool_id: Optional[str] = Field(
        default=None,
        description=(
            "Tool-call id this event was produced inside. None when "
            "produced by the parent agent. UIs use this to nest sub-"
            "agent activity (council members, critic, fusion) inside "
            "the parent tool card."
        ),
    )


class Token(_Nestable):
    """Single LLM content delta (from on_chat_model_stream)."""
    type: Literal["token"] = "token"
    content: str


class Reasoning(_Nestable):
    """Reasoning-trace delta (e.g. <think> blocks from r1-style models).
    Today we forward the raw text and let the UI's existing
    splitReasoning() / ReasoningBlock pipeline handle it."""
    type: Literal["reasoning"] = "reasoning"
    content: str


class ToolCallStart(_Nestable):
    type: Literal["tool_call_start"] = "tool_call_start"
    id: str
    name: str
    args: dict[str, Any]


class ToolCallEnd(_Nestable):
    type: Literal["tool_call_end"] = "tool_call_end"
    id: str
    result: Optional[Any] = None
    error: Optional[str] = None
    duration_ms: int


class Iteration(_Nestable):
    """New tool-loop iteration begins. Mirrors the iteration counter the
    TS SDK exposed; UIs can use it for grouping or progress hints."""
    type: Literal["iteration"] = "iteration"
    n: int


class MessageDone(BaseModel):
    type: Literal["message_done"] = "message_done"
    finish_reason: Optional[str] = None
    usage: Optional[dict[str, Any]] = None


class ErrorEvent(BaseModel):
    type: Literal["error"] = "error"
    message: str


# Discriminated union — Pydantic uses the `type` field to pick the right
# model when parsing on the UI side (TS will mirror this with a
# string-tagged union). Keep this list ordered the same way the UI
# documents the events for readability.
AgentEvent = Union[
    RunStarted,
    Token,
    Reasoning,
    ToolCallStart,
    ToolCallEnd,
    Iteration,
    MessageDone,
    ErrorEvent,
]


class ChatStreamRequest(BaseModel):
    """Body of POST /api/chat/stream — a single chat dispatch.

    The minimum viable request is `{model, messages}`; everything else
    layers on top. For per-Agent config overrides (researcher model,
    recursion limits, system prompts, …), use `agent_config` —
    those values are validated against the matching Agent's Pydantic
    config_model before they reach the graph.
    """

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "model": "claude-haiku-4-5-20251001",
                    "messages": [{"role": "user", "content": "Say hi."}],
                },
                {
                    "model": "claude-haiku-4-5-20251001",
                    "messages": [
                        {"role": "user", "content": "Research the LiteLLM release notes."}
                    ],
                    "tools": ["research"],
                    "agent_config": {
                        "main": {"recursion_limit": 10},
                        "research": {
                            "model": "claude-haiku-4-5-20251001",
                            "recursion_limit": 8,
                        },
                    },
                },
            ]
        }
    }

    model: str = Field(
        description="Model id known to the LiteLLM proxy (see GET /api/models).",
        examples=["claude-haiku-4-5-20251001", "mac/qwen3-coder"],
    )
    messages: list[dict[str, Any]] = Field(
        description="OpenAI-style message dicts: `{role, content, ...}`.",
    )
    system_prompt: Optional[str] = Field(
        default=None,
        description="Prepended to every chat-model invocation in the agent loop.",
    )
    tools: Optional[list[str]] = Field(
        default=None,
        description=(
            "Tool names from GET /api/tools. `null` or `[]` disables tools "
            "(graph degenerates to a single LLM call)."
        ),
        examples=[["web_search", "research"]],
    )
    params: Optional[dict[str, Any]] = Field(
        default=None,
        description=(
            "Legacy main-Agent sampling params (temperature, max_tokens, "
            "top_p, reasoning_effort, …). Prefer `agent_config[\"main\"]` "
            "for new code; this stays supported for backward compat."
        ),
    )
    agent_config: Optional[dict[str, dict[str, Any]]] = Field(
        default=None,
        description=(
            "Per-Agent overrides for tunables advertised on `GET /api/agents`. "
            "Each inner dict is validated through that Agent's Pydantic "
            "config_model; bogus fields raise. Backward compat: when "
            "absent, request behaves identically to today's behaviour."
        ),
        examples=[
            {
                "main": {"recursion_limit": 10, "temperature": 0.4},
                "research": {"model": "claude-haiku-4-5-20251001"},
            }
        ],
    )
